system/ee/ExpressionEngine/Addons/fluid_field/ft.fluid_field.php

```diff
@@ -436,6 +445,16 @@ private function addField($order, $group, $field_id, array $values)
 
         $fluid_field->field_data_id = $id;
         $fluid_field->save();
+
+        if (ee()->extensions->active_hook('fluid_field_after_add_field') === true) {
+            ee()->extensions->call(
+                'fluid_field_after_add_field',
+                $fluid_field,
+                $fluid_field->ChannelField->getTableName(),
+                $values,
+                $id
```

Comment:
This looks good. So far I have basic text fields working fine in Fluid and Publisher  with most of these changes, but one more small change might be necessary. I wrapped the `fluid_field-save()` call on line 447 with a conditional, so the first hook could optionally return an empty array, and if so then there isn't anything to save. I did this so I could save non-default language, otherwise they end up getting added to the native Fluid tables, which only hold the default language. For example:

```
if (ee()->extensions->active_hook('fluid_field_add_field') === true) {
            $values = ee()->extensions->call(
                'fluid_field_add_field',
                $fluid_field->ChannelField->getTableName(),
                $values,
                $fluid_field
            );
        }

        $field = ee('Model')->get('ChannelField', $field_id)->first();

        if (!empty($values)) {
            $query = ee('db');
            $query->set($values);
            $query->insert($field->getTableName());
            $id = $query->insert_id();

            $fluid_field->field_data_id = $id;
            $fluid_field->save();
        }

        if (ee()->extensions->active_hook('fluid_field_after_add_field') === true) {
            ee()->extensions->call(
                'fluid_field_after_add_field',
                $fluid_field,
                $fluid_field->ChannelField->getTableName(),
                $values,
                $id ?? 0
            );
        }
```

The conditional and the null check to default $id to 0 are the differences. I did something similar to the updateField method too:

```
if (ee()->extensions->active_hook('fluid_field_update_field') === true) {
            $values = ee()->extensions->call(
                'fluid_field_update_field',
                $fluid_field,
                $fluid_field->ChannelField->getTableName(),
                $values
            );
        }

        $fluid_field->field_group_id = array_key_exists('id', $group) ? $group['id'] : null;
        $fluid_field->group = array_key_exists('order', $group) ? $group['order'] : null;
        $fluid_field->order = $order;
        $fluid_field->save();

        if (!empty($values)) {
            $query = ee('db');
            $query->set($values);
            $query->where('id', $fluid_field->field_data_id);
            $query->update($fluid_field->ChannelField->getTableName());
        }

        if (ee()->extensions->active_hook('fluid_field_after_update_field') === true) {
            ee()->extensions->call(
                'fluid_field_after_update_field',
                $fluid_field,
                $fluid_field->ChannelField->getTableName(),
                $values
            );
        }
```

Even without the hook changes the `if (!empty()` check on those arrays could be a nice safety mechanism.